### DISCUSSION TITLE
Access translation through dot notation?

### DISCUSSION THREAD

**Discussion by gterras** [ivanhofer/typesafe-i18n#504] [Q&A] [answered]
Hi, with the following translation (svelte plugin): 

```js
const en: BaseTranslation = {
	LINE: {
		FOO: 'bar'
	},
}
``` 

Is it possible to access it dynamically from a `"LINE.FOO"` string? 

Goal is to declare base objects like 

```js
const field = {
  label: 'LINE.FOO'
}
```

and access it dynamically from a generic component 

```js
<script>
export let field
</script>
{$LL[field.label]()}
```

Thanks

**Comment by ivanhofer** [comment, marked as answer]
Hi @gterras you have the possibility to access it like you would do with any regular object:

```ts
$LL['LINE']['FOO']()

// or with dynamic keys
const part1 = 'LINE' as const
const part2 = 'FOO' as const
$LL[part1][part2]()
```

Accessing the object via a string split by `.` is not supported out of the box.
You would need to write a custom function in order to do that. It will probably require some additional effort if you want to keep the typesafety in place.

**Comment by gterras** [comment]
Thanks for your answer that's what I was afraid of, I'll try to find another way to inject the category or switch to flat object structure if not easily feasible since I don't want to manually handle this kind of logic.

**Reply by ivanhofer** [reply]
I have hacked together a solution that should work for you:

```ts
type GenNode<K extends string | number, Root extends boolean> = Root extends true
	? `${K}`
	: `.${K}` | (K extends number
		? `[${K}]` | `.[${K}]`
		: never);

type ObjectKeyPaths<
	T extends object,
	Root extends boolean = true,
	K extends keyof T = keyof T
> = K extends string | number
	? GenNode<K, Root> | (T[K] extends object
		? `${GenNode<K, Root>}${ObjectKeyPaths<T[K], false>}`
		: never)
	: never;

type GetT<Obj, Path extends string> = Path extends ''
	? Obj
	: Path extends `${infer Key}.${infer Rest}`
	? Key extends keyof Obj
	? GetT<Obj[Key], Rest>
	: never
	: Path extends keyof Obj
	? GetT<Obj[Path], ''>
	: never

const getByPath = <Obj extends BaseTranslation, Path extends ObjectKeyPaths<Obj>>(obj: Obj, path: Path): GetT<Obj, Path> => {
	let o: any = obj
	path.split('.').forEach(key => o = o[key])
	return o
}

// usage
getByPath($LL, 'LINE.FOO')
```

**Reply by gterras** [reply]
That looks very clever thanks a lot I will try this shortly!

**Reply by gterras** [reply]
It works really nicely and elegantly thank you. I just couldn't figure out how to properly type the string declaration `type DottedPath = ObjectKeyPaths<BaseTranslation, true, keyof BaseTranslation>`

**Reply by ivanhofer** [reply]
`type DottedPath = ObjectKeyPaths<Translation>` should be enough

**Reply by gterras** [reply]
That's perfect thank you!

**Reply by ivanhofer** [reply]
you're welcome
